File: leetcode/710.黑名单中的随机数.go
package main

import "math/rand"

/*
 * @lc app=leetcode.cn id=710 lang=golang
 *
 * [710] 黑名单中的随机数
 */

// @lc code=start
type Solution struct {
	bound int
	data  map[int]int
}

func Constructor(n int, blacklist []int) Solution {
	m := len(blacklist)
	bound := n - m
	start := n - m
	blackMap := map[int]bool{}
	for _, b := range blacklist {
		if b >= bound {
			blackMap[b] = true
		}
	}

	data := make(map[int]int, m-len(blackMap))
	for _, b := range blacklist {
		if b < bound {
			for blackMap[start] {
				start++
			}
			data[b] = start
			start++
		}
	}
	return Solution{bound, data}
}

func (this *Solution) Pick() int {
	res := rand.Intn(this.bound)
	if this.data[res] > 0 {
		return this.data[res]
	}
	return res
}

/**
 * Your Solution object will be instantiated and called as such:
 * obj := Constructor(n, blacklist);
 * param_1 := obj.Pick();
 */
// @lc code=end

File: leetcode/710.黑名单中的随机数.cpp
#include <map>
#include <set>
#include <vector>
using namespace std;
/*
 * @lc app=leetcode.cn id=710 lang=cpp
 *
 * [710] 黑名单中的随机数
 */

// @lc code=start
class Solution
{
private:
    map<int, int> data;
    int bound;

public:
    Solution(int n, vector<int> &blacklist)
    {
        int m = blacklist.size();
        bound = n - m;
        int start = bound;
        set<int> blackSet;
        for (int b : blacklist)
        {
            if (b >= bound)
            {
                blackSet.emplace(b);
            }
        }
        for (auto &&b : blacklist)
        {
            if (b < bound)
            {
                while (blackSet.count(start))
                    start++;
                data[b] = start;
                start++;
            }
        }
    }

    int pick()
    {
        int res = rand() % bound;
        return data.count(res) ? data[res] : res;
    }
};

/**
 * Your Solution object will be instantiated and called as such:
 * Solution* obj = new Solution(n, blacklist);
 * int param_1 = obj->pick();
 */
// @lc code=end
